File: LeetCode/0666.Path-Sum-Iv/Path-Sum-Iv.java
class Solution {
    int sum = 0;
    Map<Integer, Integer> map = new HashMap<>();
    public int pathSum(int[] nums) {
        for(int num : nums){
            int key = num / 10;
            int val = num % 10;
            map.put(key, val);
        }
        traverse(nums[0] / 10, 0);
        return sum;
    }
    public void traverse(int key, int pre){
        int level = key / 10;
        int pos = key % 10;
        int left = (level + 1) * 10 + pos * 2 - 1;
        int right = (level + 1) * 10 + pos * 2;
        int curSum = pre + map.get(key);
        if(!map.containsKey(left) && !map.containsKey(right)){
            sum += curSum;
            return;
        }
        if(map.containsKey(left)) traverse(left, curSum);
        if(map.containsKey(right)) traverse(right, curSum);
        
    }
}

File: LeetCode/0666.Path-Sum-Iv/Path-Sum-Iv.py
class Solution(object):
    def pathSum(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        #    0
        #  1   2      2^1 - 2 + n
        # 3 4 5 6     2^2 - 2 + n
        
        n = len(nums)
        if n == 1: return nums[0] % 10
        tree = [-1] * 32
        for num in nums:
            d, p, v = num // 100, (num % 100) // 10, num % 10
            tree[2 ** (d - 1) - 2 + p] = v
            
        self.res = 0
            
        def dfs(idx, cur):
            if tree[idx] == -1: return 0
            cur += tree[idx]
            if idx >= 15 or (tree[2 * idx + 1] == -1 and tree[2 * idx + 2] == -1):
                self.res += cur
                return
                
            dfs(2 * idx + 1, cur)
            dfs(2 * idx + 2, cur)
            
        dfs(0, 0)    
            
        return self.res
            

        
